com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -588,216 +1010,32 @@ private void Initialize(bool server)
-                if (playerPrefabNetworkObject != null)
-                {
-                    //In the event there is no NetworkPrefab entry (i.e. no override for default player prefab)
-                    if (!NetworkConfig.NetworkPrefabOverrideLinks.ContainsKey(playerPrefabNetworkObject
-                        .GlobalObjectIdHash))
-                    {
-                        //Then add a new entry for the player prefab
-                        var playerNetworkPrefab = new NetworkPrefab();
-                        playerNetworkPrefab.Prefab = NetworkConfig.PlayerPrefab;
-                        NetworkConfig.NetworkPrefabs.Insert(0, playerNetworkPrefab);
-                        NetworkConfig.NetworkPrefabOverrideLinks.Add(playerPrefabNetworkObject.GlobalObjectIdHash,
-                            playerNetworkPrefab);
-                    }
-                }
-                else
-                {
-                    // Provide the name of the prefab with issues so the user can more easily find the prefab and fix it
-                    Debug.LogError($"{nameof(NetworkConfig.PlayerPrefab)} (\"{NetworkConfig.PlayerPrefab.name}\") has no NetworkObject assigned to it!.");
-                }
-            }
-
-            // Clear out anything that is invalid or not used (for invalid entries we already logged warnings to the user earlier)
-            // Iterate backwards so indices don't shift as we remove
-            for (int i = removeEmptyPrefabs.Count - 1; i >= 0; i--)
-            {
-                NetworkConfig.NetworkPrefabs.RemoveAt(removeEmptyPrefabs[i]);
-            }
-
-            removeEmptyPrefabs.Clear();
-
             NetworkConfig.NetworkTransport.OnTransportEvent += HandleRawTransportPoll;
 
             NetworkConfig.NetworkTransport.Initialize(this);
         }
 
         /// <summary>
         /// Starts a server
+        /// Returns true if the Server *has not yet failed* to start
+        /// There are several steps to connecting - a return value of "true" here does not indicate
+        /// the connection is completed.
+        ///
+        /// When the connection has successfully been established the onReady callback (or any callback
+        /// registered with OnReadyCallback) will be invoked.
+        ///
+        /// If any step of the connection process fails, the onFailure callback (or any callback
+        /// registered with OnStartupFailedCallback) will be invoked with a value indicating the
+        /// reason that startup failed.
+        ///
+        /// These are the most reliable ways to check for startup success and failure - the
+        /// boolean return value exists for legacy reasons but is not a reliable indicator
+        /// of startup success.
         /// </summary>
-        public bool StartServer()
+        public bool StartServer(Action onReady = null, Action<StartupFailureReason> onFailure = null)
```

Comment:
Note to myself, should probably unregister these functions automatically at shutdown.